common/serde-reflection/src/format.rs

```diff
@@ -37,54 +37,67 @@ pub enum Format {
     Bytes,
 
-    /// The format of Option<T>
+    /// The format of `Option<T>`.
     Option(Box<Format>),
-    /// E.g. the format of Vec<Foo>
+    /// A sequence, e.g. the format of `Vec<Foo>`.
     Seq(Box<Format>),
-    /// E.g. the format of BTreeMap<K, V>
+    /// A map, e.g. the format of `BTreeMap<K, V>`.
     #[serde(rename_all = "UPPERCASE")]
     Map {
         key: Box<Format>,
         value: Box<Format>,
     },
 
-    /// E.g. the format of (Foo, Bar)
+    /// A tuple, e.g. the format of `(Foo, Bar)`.
     Tuple(Vec<Format>),
-    /// Alias for (Foo, .... Foo)
-    /// E.g. the format of [Foo; N]
+    /// Alias for `(Foo, ... Foo)`.
+    /// E.g. the format of `[Foo; N]`.
     #[serde(rename_all = "UPPERCASE")]
     TupleArray {
         content: Box<Format>,
         size: usize,
     },
+}
 
-    /// The format of an empty struct
+/// Serde-based serialization format for named "container" types.
+/// In Rust, those are enums and structs.
+#[derive(Serialize, Deserialize, Debug, Eq, Clone, PartialEq)]
+#[serde(rename_all = "UPPERCASE")]
+pub enum ContainerFormat {
+    /// An empty struct, e.g. `struct A`.
     UnitStruct,
-    /// The format of a struct with a unique unnamed field, e.g. struct A (u16)
+    /// A struct with a single unnamed parameter, e.g. `struct A(u16)`
     NewTypeStruct(Box<Format>),
-    /// The format of a struct with several unnamed fields, e.g. struct A (u16, u32)
+    /// A struct with several unnamed parameters, e.g. `struct A(u16, u32)`
     TupleStruct(Vec<Format>),
-    /// The format of a struct with named fields, e.g. struct A { a: Foo }
+    /// A struct with named parameters, e.g. `struct A { a: Foo }`.
     Struct(Vec<Named<Format>>),
-    /// The format of enum containers
-    Variant(BTreeMap<u32, Named<VariantFormat>>),
+    /// An enum, that is, an enumeration of variants.
+    /// Each variant has a unique name and index within the enum.
+    Enum(BTreeMap<u32, Named<VariantFormat>>),
 }
 
-#[derive(Debug, Default, Eq, PartialEq)]
+#[derive(Debug, Clone, Default, Eq, PartialEq)]
 /// A named value.
-/// Used for named fields or variant cases.
+/// Used for named parameters or variants.
 pub struct Named<T> {
     pub name: String,
     pub value: T,
 }
 
-#[derive(Serialize, Deserialize, Debug, Eq, PartialEq)]
+#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq)]
 #[serde(rename_all = "UPPERCASE")]
-/// Description of a variant case.
+/// Description of a variant in an enum.
 pub enum VariantFormat {
+    /// A variant whose format is yet unknown.
+    Unknown,
+    /// A variant without parameters, e.g. `A` in `enum X { A }`
     Unit,
+    /// A variant with a single unnamed parameter, e.g. `A` in `enum X { A(u16) }`
     NewType(Box<Format>),
+    /// A struct with several unnamed parameters, e.g. `A` in `enum X { A(u16, u32) }`
     Tuple(Vec<Format>),
+    /// A struct with named parameters, e.g. `A` in `enum X { A { a: Foo } }`
     Struct(Vec<Named<Format>>),
 }
 
```

Comment:
`merge` is a (linear) term unification algorithm. We are trying to find the value of the different `Unknown`s that make `v1` and `v2` equal. If don't have the same structure, it will fail.

Reply:
For performance reasons (and simplicity), we (optimistically) destroy v2 to update v1 in place.